llvm/include/llvm/IR/DataLayout.h

```diff
@@ -245,6 +246,7 @@ class DataLayout {
   unsigned getDefaultGlobalsAddressSpace() const {
     return DefaultGlobalsAddrSpace;
   }
+  unsigned getFlatAddressSpace() const { return FlatAddressSpace; }
```

Comment:
> The absence of this specification indicates the target does not have such a flat address space to optimize away.

Yet this isn't an optional? Is the expectation that callers check against a magic ~0U sentinel? (Is there precedent for that in DataLayout?)

Reply:
It is optional from a target's perspective, but the expectation is indeed that to check against `~0U` because it represents invalid address space. I will put a comment there.

Reply:
I think that for general ergonomics making this an `optional` is preferable to magic constant + comment. That'd unambiguously convey absence, and it's what optional is for. Whilst incredibly unlikely, it is currently not impossible for some future target do decide that AS UINT32_MAX carries semantic meaning, and run into trouble around this.

Reply:
I added an enum such that all the checks can be against `DataLayout::AS_INVALID`.

Reply:
I don't think this is a good API - returning something that looks valid (uint32_t) but might be incorrect at runtime without forcing users to check it.

Storing it as ~0 internally is fine but IMO the API should return `optional` and map ~0 to `std::nullopt`.